### PR TITLE
Add TransferHooks, Budget fix and upgrade to Polkadot v0.9.34

### PR DESCRIPTION
**Pull request by HashWarlock**
### TransferHooks implementation
For implementations downstream of the RMRK pallet there needs to be an ability to add custom logic in the case that there are multiple pallets using the RMRK NFT pallet, but want to keep the business logic mutually exclusive. This implementation allows for a trait called `TransferHooks` to be implemented with 2 functions `pre_check` and `post_transfer` that default to `true` when not implemented.

### Budget trait fix
There are 2 problems with the current Budget trait:

NFTs are currently allowed to mint and send NFTs to a parent NFT that can meet up to the threshold of NestingBudget::get() + 1. This allows for NFTs queries like lookup_root_owner to return errors like TooManyRecursions. Having this error is okay, but the NFTs should be prevented from being minted to or sent to a parent NFT if the Budget threshold is met.

Currently the code will do a calculation on functions like burn_nft and this will perform an arithmetic subtraction at the end of the function, but this should be added as a field in the Value type & implement a function from Budget to get the budget that has been consumed and budget that is left.

## Targets
- [x] Upgrade to `polkadot-v0.9.34`
- [x] implement `TransferHooks` trait with `pre_check` and `post_transfer` functions
- [x] Update integration tests
- [x] Fix minting/sending NFTs that exceed the threshold of `NestingBudget::get()` value.
- [x] Add fields names `budget_consumed` and budget_left` to track & through errors is an overflow is detected.
- [x] Add functions in `Budget` trait to get the 2 new fields.
- [x] Fix benchmarking errors from Traits being renamed

### CODE REVIEW

**Review by Szegoo — APPROVED** [review]
LGTM.

File: traits/src/misc.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Copyright (C) 2021-2022 RMRK
+// This file is part of rmrk-substrate.
+// License: Apache 2.0 modified by RMRK, see LICENSE.md
+
+/// Trait for pre-checks and post-checks for transfers that can be implemented downstream to extend
+/// the logic of RMRK's current funcitonality.
+pub trait TransferHooks<AccountId, CollectionId, NftId> {
+	/// Check if the NFT's pre-checks and post-checks for the transfer function based on the sender,
+	/// collection_id and nft_id parameters.
```
Comment (Szegoo) [inline traits/src/misc.rs:9]:
```suggestion
	/// `collection_id` and `nft_id` parameters.
```

File: pallets/rmrk-core/src/tests.rs [hunk 2]
```diff
@@ -1229,12 +1229,22 @@ fn burn_nft_beyond_max_recursions_fails_gracefully() {
 			3,
 			AccountIdOrCollectionNftTuple::CollectionAndNftTuple(0, 2),
 		));
-		// ALICE sends NFT (0, 4) to NFT (0, 3)
+		// ALICE sends NFT (0, 4) to NFT (0, 3) fails bc there nested budget threshold is exceeded
```
Comment (Szegoo) [inline pallets/rmrk-core/src/tests.rs:1232]:
```suggestion
		// ALICE sends NFT (0, 4) to NFT (0, 3) fails bc the nested budget threshold is exceeded
```

**Comment by HashWarlock** [comment]
Closing in favor of #261

**Pull request closed without merging**